### PR TITLE
Experimenting with hybrid evaluation again.

### PR DESCRIPTION
**Pull request by romainfrancois**
It's very shaky for now, here are some details: 

before the usual loop is evaluated with `mask$eval_all_summarise(quo)` this first calls `hybrid_eval_summarise()` which might return: 

 - `NULL` : hybrid evaluation failed, just proceed as normal
 - a list that may contain:
     - `$chunks`: the result chunks, this is what `mask$eval_all_summarise(quo)` would give
     - `$sizes`: same as `list_sizes($chunks)`
     - `$results`: same as `vec_c(!!!$chunks)`

Things are made complicated because `summarise()` no longer restricts output size to `1` so we can't just have `$results` but perhaps there  could be a way to make `$chunks` lazy as it might not be needed by follow up operations. 

Then there's `hybrid_eval_summarise()` which for now takes `mask` and `quo` as its arguments and uses `mask$current_vars()` and `mask$resolve()`. 

This leaves chopping responsibility to the DataMask class so that's good. 

At the moment, `hybrid_eval_summarise` looks at the call, but this part is even more sketchy

### CODE REVIEW

**Comment by romainfrancois** [comment]
Making some progress here. I'll start writing a document about the approach. 

This tries to separate the concerns of hybridability and the actual calculation. `hybrid_eval_summarise()` is mostly in charge of checking if `expr` can be handled hybridly. 

And then `grouped_mean()` has no "magic", it would e.g. live in `funs::` and be passed a list of vectors as `x`: the current implementation is just to play with the mechanism.  

```r
grouped_mean <- function(x, na.rm = FALSE) {
  # TODO: rewrite in C++
  if (na.rm) {
    x <- map(x, function(.x) .x[!is.na(.x)])
  }
  res <- map(x, function(.x) .Internal(mean(.x)))

  vec_c(!!!res)
}
```

I've added the `hybrid()` function as a mechanism so that we could write `grouped_*()` functions independently, e.g. 

``` r
library(dplyr, warn.conflicts = FALSE)

grouped_sum <- function(x) {
  purrr::map_dbl(x, sum)
}

data.frame(x = c(1, 1, 5), y = 1:3) %>% 
  group_by(x) %>% 
  summarise(n = hybrid(grouped_sum(x)))
#> # A tibble: 2 x 2
#>       x     n
#> * <dbl> <dbl>
#> 1     1     2
#> 2     5     5
```

In short, this says `grouped_sum(x)` is called in a context where `x` is the chops of `x`, i.e. `list(c(1,1), c(5))`.

**Comment by romainfrancois** [comment]
Closing this now, will reboot this idea after #5524 is merged
